File: frontend/src/lib/api/invites.ts
import { apiFetch } from "./client";

export interface InvitePreview {
  email: string;
  role: string;
  workspace: { id: string; name: string; slug: string };
}

export interface CreateInviteResponse {
  id: string;
  email: string;
  role: string;
  token: string;
  inviteUrl: string;
}

export interface AcceptInviteSignupResponse {
  user: { id: string; email: string; fullName: string };
  accessToken: string;
  workspace: { id: string; name: string; slug: string };
  role: string;
  flow: "invitee";
}

export function createInvite(
  token: string,
  workspaceId: string,
  email: string,
  role: string
) {
  return apiFetch<CreateInviteResponse>(
    `/workspaces/${workspaceId}/invites`,
    {
      method: "POST",
      token,
      body: JSON.stringify({ email, role }),
    }
  );
}

export function getInvitePreview(inviteToken: string) {
  return apiFetch<InvitePreview>(`/invites/${inviteToken}`);
}

export function acceptInviteSignup(
  inviteToken: string,
  fullName: string,
  password: string
) {
  return apiFetch<AcceptInviteSignupResponse>(
    `/invites/${inviteToken}/accept-signup`,
    {
      method: "POST",
      body: JSON.stringify({ fullName, password }),
    }
  );
}

export function acceptInvite(token: string, inviteToken: string) {
  return apiFetch<{
    workspace: { id: string; name: string; slug: string };
    role: string;
    flow: string;
  }>(`/invites/${inviteToken}/accept`, { method: "POST", token });
}

File: backend-py/app/api/v1/invites.py
from fastapi import APIRouter, Request, Response, status

from app.api.cookies import set_refresh_cookie
from app.core.rate_limit import throttle
from app.deps.auth import CurrentUserDep, DbSession
from app.schemas.workspace import AcceptInviteSignupBody
from app.services import invite_service

router = APIRouter(prefix="/invites", tags=["invites"])


@router.get("/{token}")
async def preview_invite(request: Request, token: str, session: DbSession):
    """Preview invite (public)."""
    await throttle(request, scope="invites:preview", ip_limit=60, window_seconds=60)
    return await invite_service.get_invite_by_token(session, token)


@router.post("/{token}/accept")
async def accept_invite(
    request: Request,
    token: str,
    session: DbSession,
    user: CurrentUserDep,
):
    """Accept invite while logged in."""
    await throttle(request, scope="invites:accept", ip_limit=30, window_seconds=60)
    return await invite_service.accept_invite_for_user(session, token, user.id)


@router.post("/{token}/accept-signup", status_code=status.HTTP_201_CREATED)
async def accept_invite_signup(
    request: Request,
    token: str,
    body: AcceptInviteSignupBody,
    response: Response,
    session: DbSession,
):
    """Accept invite and create a new account."""
    await throttle(
        request, scope="invites:accept_signup", ip_limit=20, window_seconds=60
    )
    result = await invite_service.accept_invite_with_signup(
        session, token, body.full_name, body.password
    )
    set_refresh_cookie(response, result.pop("refreshToken"))
    return {
        "user": result["user"],
        "accessToken": result["accessToken"],
        "workspace": result["workspace"],
        "role": result["role"],
        "flow": "invitee",
    }
